---
title: "Numerical Methods for PDEs Mini-Project 2"
author:
- Ashleigh Trinh
- Paper by Andrea Beaghton, Pantelis John Beaghton & Austin Burt
date: ''
output:
  html_document:
    toc: yes
  html_notebook: default
  toc: yes
  word_document:
    toc: yes
subtitle: 'Brief mathematical analysis of: Gene drive through a landscape: Reaction-diffusion
  model of population suppression and elimination by a sex ratio distorter'
---

For this project I will be looking at a system from the paper:
Beaghton, A., Beaghton, P. J., and Burt, A. (2016). Gene drive through a landscape: Reaction-diffusion models of population suppression and elimination by a sex ratio distorter. _Theoretical Population Biology_, 108, 51-69.

Note: all references made to equation and section numbers follows the equations within the lecture notes, Numerical Methods for PDEs Lecture Notes (Spring 2017).

## Section 1 - Biological Summary

The paper looks at the driving-Y chromosome which is a sex ratio distorter, such that it transmits to more than 90% of progeny. This in theory, leads to the majority of its offspring being male. However, this is rarely reported in nature due to the presence of resistant alleles, further research has found that Y drive occurs due to the disruption of the development of X-bearing sperm during male meiosis.

Beaghton et al. (2016) uses the example of _Anopheles gambiae_ mosquito populations where there is a synthetic gene constructed to 'break' the X chromosome during male meiosis. So with this synthetic gene there is a 95% chance of transmitting the Y chromosome to its progeny implying a synthetic driving-Y.

## Section 2 - Introduction to Model System

This system I am considering describes the introduction of driving-Y males into a one life stage system of wild-type males and females to model the spatial spread of the invasive species through the region. In the model, each birth is immediately preceeded by a female mating with a driving-Y or wild-type male chosen at random according to the local (in time and space) proportion of each in the total male population:
$$
\frac{\partial{H}}{\partial{t}} = D_{H}\frac{\partial^2{H}}{\partial{x^2}} + 2\lambda F\left[\frac{mH}{M+H}\right] - (\mu + \gamma N)H \\
\frac{\partial{F}}{\partial{t}} = D_{F}\frac{\partial^2{F}}{\partial{x^2}} + 2\lambda F\left[\frac{M}{2(M+H)} + \frac{(1-m)H}{M+H}\right] - (\mu + \gamma N)F \\
\frac{\partial{M}}{\partial{t}} = D_{M}\frac{\partial^2{M}}{\partial{x^2}} + 2\lambda F \left[\frac{M}{2(M+H)}\right] - (\mu + \gamma N)M
$$
where the variables are $H(x,t)$, $F(x,t)$, and $M(x,t)$ are population densities of driving-Y males, females, and males at position $x$ at time $t$, respectively. Also $N(x,t) = H(x,t) + F(x,t) + M(x,t)$ which is the total population at position $x$ at time $t$. 

The constant parameters considered in the paper for _An. gambiae_ are as follows:

$D_i$ - diffusion coefficient of adult type $U_i$,

$m$ - fraction of progeny of driving-Y males that inherit the driving-Y,

$\lambda$ - fecundity (female eggs per female),

$\mu$ - adult density-independent mortality coefficient,

$\gamma$ - adult density-dependent mortality coefficient,

$\rho$ - density of release of driving-Y at $t = 0$.

This system of equations is subject to the initial condition(s):
$$
H(x,0) = 
\begin{cases}
\rho,& \text{if } -a \leq x \leq a \\
0,& \text{if } x > |a|.
\end{cases} \\ \\
F(x,0)=M(x,0)=\frac{N_0}{2} \\
\text{where } N_0 = \frac{\lambda - \mu}{\gamma}
$$

## Section 3 - Method

To begin with I will be implementing the forward difference method for this set of reaction-diffusion equations. 

Each of the equations contains a linear second-order derivative term which will be discretised as shown in Eq. (2.7) using the central difference method:
$$
\frac{\partial^2{H}}{\partial{x^2}}(x_k,t_j) = \frac{H_{k+1,j} - 2H_{k,j} + H_{k-1,j}}{h^2} - \mathcal{O}(h^2)
$$
and similarly for $\partial^2F/\partial{x^2}$ and $\partial^2M/\partial{x^2}$.

The system's time derivatives are discretised by the forward difference formula:
$$
\frac{\partial{H}}{\partial{t}}(x_k,t_j) = \frac{H_{k,j+1} - H_{k.j}}{\tau} + \mathcal{O}(\tau)
$$
and similarly for $\partial{F}/\partial{t}$ and $\partial{M}/\partial{t}$.

Compiling the above, I get the following explicit method:
$$
\begin{split}
H_{k,j+1} = H_{k,j} + \tau \Bigg(
&\frac{D}{h^2}\left(H_{k+1,j}-2H_{k,j}+H_{k-1,j}\right) \\
&+ 2\lambda F_{k,j} \left[\frac{mH_{k,j}}{M_{k,j} + H_{k,j}}\right] - \left(\mu + \gamma N_{k,j}\right)H_{k,j}
\Bigg) \\
F_{k,j+1} = F_{k,j} + \tau \Bigg(
&\frac{D}{h^2}\left(F_{k+1,j}-2F_{k,j}+F_{k-1,j}\right) \\
&+ 2\lambda F_{k,j} \left[\frac{M_{k,j}}{2\left(M_{k,j} + H_{k,j}\right)} + \frac{(1-m)H_{k,j}}{M_{k,j} + H_{k,j}}\right] - \left(\mu + \gamma N_{k,j}\right)F_{k,j}
\Bigg) \\ 
M_{k,j+1} = M_{k,j} +\tau \Bigg(
&\frac{D}{h^2}\left(M_{k+1,j}-2M_{k,j}+M_{k-1,j}\right) \\
&+ 2\lambda F_{k,j} \left[\frac{M_{k,j}}{2\left(M_{k,j} + H_{k,j}\right)}\right] - \left(\mu + \gamma N_{k,j}\right)M_{k,j}
\Bigg)
\end{split}
$$
This method has a truncation error of $\mathcal{O}(\tau + h^2)$.


## Section 4 - Boundary Conditions

The paper did not define specific boundary conditions hence I choose to impose Neumann boundary conditions. The paper implies that $H$ is confined to an area $-a \leq x \leq a$ however $F$ and $M$ are not constrained by $a$ so I set the plot size in one direction to $\frac{Q}{2}$ where $h\frac{Q}{2} > a$ and $h$ is the length size of the grid points, so I define the boundary conditions as:
$$
\begin{align}
\frac{\partial{H}}{\partial{x}}(-a,t) &= \frac{\partial{H}}{\partial{x}}(a,t) = 0 \\ \\ \frac{\partial{F}}{\partial{x}}\left(\frac{-Q}{2},t\right) = \frac{\partial{F}}{\partial{x}}\left(\frac{Q}{2},t\right) &= \frac{\partial{M}}{\partial{x}}\left(\frac{-Q}{2},t\right) = \frac{\partial{M}}{\partial{x}}\left(\frac{Q}{2},t\right) = 0
\end{align}
$$
Note that I set the boundaries as $\pm \frac{Q}{2}$ since my vector length will be $Q$ but I want it to span the positive and negative space. For more clarity refer to `Chunk 2`.

To maintain the truncation error of $\mathcal{O}(h^2)$ I implement the scheme from section 2.6 by introducing the points $x_{-\frac{Q}{2}-1}$ and $x_{\frac{Q}{2}+1}$ and using central-difference formula for the derivatives to translate the boundary conditions into expressions for the values of the solution at these points:
$$
\begin{align}
F_{\frac{Q}{2}+1,j} &= F_{\frac{Q}{2}-1,j},&~~~  F_{-\frac{Q}{2}-1,j} &= F_{-\frac{Q}{2}+1,j},\\
M_{\frac{Q}{2}+1,j} &= M_{\frac{Q}{2}-1,j},&~~~  M_{-\frac{Q}{2}-1,j} &= M_{-\frac{Q}{2}+1,j}.
\end{align}
$$
The boundary conditions for $H$ are different since it is bound by $a$ which gives us the points:
$$
H_{a+1,j} = H_{a-1,j},~~~~~~~~~~~~~ H_{-a-1,j} = H_{-a+1,j}
$$
Since the grid of points are referenced as ranges from $0$ to $Q$ where $-\frac{Q}{2} \implies [0,~]$ and $\frac{Q}{2} \implies [Q,~]$ which gives us a result analogous to that shown in Lab 11 for $F$ and $M$:
$$
\begin{align}
F_{Q+1,j} &= F_{Q-1,j},~~ &F_{-1,j} &= F_{1,j}, \\
M_{Q+1,j} &= M_{Q-1,j},~~ &M_{-1,j} &= M_{1,j}.
\end{align}
$$
This means that the above finite difference equations will work with $k = 0$ and $k = Q$.

The expressions can be converted into vectors, which is easier to translate into R code:
$$
\vec{F_{j}} =
\begin{pmatrix} F_{0,j} \\ F_{1,j} \\ \vdots \\ F_{Q-1,j} \\ F_{Q,j} \end{pmatrix}, ~~~
\vec{F}^+_j =
\begin{pmatrix} F_{1,j} \\ F_{2,j} \\ \vdots \\ F_{Q,j} \\ F_{Q-1,j} \end{pmatrix}, ~~~
\vec{F}^-_j =
\begin{pmatrix} F_{1,j} \\ F_{0,j} \\ \vdots \\ F_{Q-2,j} \\ F_{Q-1,j} \end{pmatrix}.
$$
and similarly for $M$. 

I translate $H$ differently to the vectors:
$$
\vec{H_{j}} =
\begin{pmatrix} H_{0,j} \\ H_{1,j} \\ \vdots \\ H_{\frac{Q}{2}-a-1,j} \\ H_{\frac{Q}{2}-a,j} \\ \vdots \\ H_{\frac{Q}{2}+a,j} \\ H_{\frac{Q}{2}+a+1,j} \\ \vdots \\ H_{Q-1,j} \\ H_{Q,j} \end{pmatrix}, ~~~
\vec{H}^+_j =
\begin{pmatrix} H_{1,j} \\ H_{2,j} \\ \vdots \\  H_{\frac{Q}{2}-a-1,j} \\ H_{\frac{Q}{2}-a,j} \\ \vdots \\ H_{\frac{Q}{2}+a,j} \\ H_{\frac{Q}{2}+a-1,j} \\ \vdots \\ H_{Q-2,j} \\ H_{Q-1,j} \end{pmatrix}, ~~~
\vec{H}^-_j =
\begin{pmatrix} H_{1,j} \\ H_{2,j} \\ \vdots \\  H_{\frac{Q}{2}-a+1,j} \\ H_{\frac{Q}{2}-a,j} \\ \vdots \\ H_{\frac{Q}{2}+a,j} \\ H_{\frac{Q}{2}+a+1,j} \\ \vdots \\ H_{Q-2,j} \\ H_{Q-1,j} \end{pmatrix}.
$$
The issue I find here is how to convert this into coherent and useful code in `R`.

## Section 5 - Coding the Method

First, I define the system parameters:
```{r, Specifying the System Parameters}
D <- 0.1
m <- 0.95
lambda <- 0.72
gamma <- 0   # For An. gambiae, density-dependent death occurs mainly at the juvenile stage.
mu <- 0.12
N0 <- (lambda - mu)
rho <- 0.01*N0
```

The model in the paper uses positive and negative space to show spread of the gene from the origin of a confined area with boudnary $a$. To show this I want my grid points to reflect all points outward from the origin. I use `rev()` to get the vertical mirror image of the vector of positive grid points resulting in the negative grid points. 
I specify the grid points for $x_k = kh$ for $k = 0, ..., Q$
```{r, Specifying the Grid Points}
Q <- 300   # Number of points after the initial state
a <- 2.5*round(sqrt(lambda/D))   # The absolute value of the maximum length of the confining area
h <- 0.1   # Length size of grid points
A <- a/h   # Number of entries within the constraint x <= abs(a)
x <- h*(1:(Q/2))   # Vector of positive non-origin grid points
xflip <- -1*rev(x)   # Vector of negataive non-origin grid points
xLong <- c(xflip, 0, x)   # Vector of positive and negative grid points
X <- (Q/2)+1   # Midpoint of xLong (where 0 lies)
```

and the time steps for the grid points $t_j = \tau j$ for $j = 0, ..., P$ where $\tau$ is the step size
```{r, Specifying the Time Steps}
P <- 500   # Number of time points after the initial state
T <- 10   # Final time
tau <- T/P  
t <- tau*(0:P)   # Vector of time steps
```

I create three matrices to hold the solutions:
```{r, Solution Matrices}
wH <- matrix(0, nrow = Q+1, ncol = P+1)
wF <- matrix(0, nrow = Q+1, ncol = P+1)
wM <- matrix(0, nrow = Q+1, ncol = P+1)
```

I implement the initial conditions given by the paper, as defined above:
```{r, Setting the Initial Conditions}
Hnt <- ifelse(xflip>=-a, rho, 0)
Hnb <- ifelse(x<=abs(a), rho, 0)
Hn <- c(Hnt, rho, Hnb)
Fn <- N0/2 + 0*xLong
Mn <- N0/2 + 0*xLong
wH[ ,1] <- Hn
wF[ ,1] <- Fn
wM[ ,1] <- Mn
```

Now I can run a `for` loop to apply the forward-difference formulae for the system above:
```{r, Forward-difference System}
for (j in 1:P) {
    
    H <- Hn
    Hp <- c(H[1:(X+A)], H[X+A-1], H[(X+A+2):(Q+1)])
    Hm <- c(H[1:(X-A-2)], H[X-A+1], H[(X-A):(Q+1)])
    
    F <- Fn
    Fp <- c(F[2:(Q+1)], F[Q])
    Fm <- c(F[2], F[1:Q])
    
    M <- Mn
    Mp <- c(M[2:(Q+1)], M[Q])
    Mm <- c(M[2], M[1:Q])
    
    N <- H + F + M
    
    Hn <- H + tau * (D/h^2 * (Hp - 2*H + Hm) + 2*lambda*F*(m*H/(M + H)) - (mu + gamma*N)*H
    )
    
    Fn <- F + tau * (D/h^2 * (Fp - 2*F + Fm) + 2*lambda*F*(M/(2*(M + H)) + (1 - m)*H/(M + H)) - (mu + gamma*N)*F
    )
    
    Mn <- M + tau * (D/h^2 * (Mp - 2*M + Mm) + 2*lambda*F*(M/(2*(M + H))) - (mu + gamma*N)*M
    )
    
  wH[ ,j+1] <- signif(Hn,4)
  wF[ ,j+1] <- signif(Fn,4)
  wM[ ,j+1] <- signif(Mn,4)
}
```

To explore the parameter space, I shall encapsulate the code we used above in a function.
```{r, Creating a Function}
genedrive <- function(deltaH=function(x) 0*x, deltaF=function(x) 0*x, deltaM=function(x) 0*x, 
                      D=0.1, mu=0.07, lambda=1, gamma=0.01, m=0.95, N0=(lambda - mu), rho=0.01*N0, a=2.5*round(sqrt(lambda/D)), 
                      L=30, Q=300, T=10, P=500) {
  
  h <- L/Q
  x <- h*(1:(Q/2))
  xflip <- -1*rev(x)   # Vector of negative non-origin grid points
  xLong <- c(xflip, 0, x)   # Vector of positive and negative grid points
  A <- a/h
  X <- (Q/2)+1   # Midpoint of xLong (where 0 lies)
  
  tau <- T/P
  t <- tau*(0:P)
  
  wH <- matrix(0, nrow = Q+1, ncol = P+1)
  wF <- matrix(0, nrow = Q+1, ncol = P+1)
  wM <- matrix(0, nrow = Q+1, ncol = P+1)
  
  Hnt <- ifelse(xflip>=-a, rho, 0)
  Hnb <- ifelse(x<=abs(a), rho, 0)
  Hn <- c(Hnt, rho, Hnb)
  Fn <- N0/2 + 0*xLong
  Mn <- N0/2 + 0*xLong
  
  wH[ ,1] <- Hn
  wF[ ,1] <- Fn
  wM[ ,1] <- Mn
  
  for (j in 1:P) {
    
    H <- Hn
    Hp <- c(H[1:(X+A)], H[X+A-1], H[(X+A+2):(Q+1)])
    Hm <- c(H[1:(X-A-2)], H[X-A+1], H[(X-A):(Q+1)])
    
    F <- Fn
    Fp <- c(F[2:(Q+1)], F[Q])
    Fm <- c(F[2], F[1:Q])
    
    M <- Mn
    Mp <- c(M[2:(Q+1)], M[Q])
    Mm <- c(M[2], M[1:Q])
    
    N <- H + F + M
    
    Hn <- H + tau * (D/h^2 * (Hp - 2*H + Hm) + 2*lambda*F*(m*H/(M + H)) - (mu + gamma*N)*H
    )
    
    Fn <- F + tau * (D/h^2 * (Fp - 2*F + Fm) + 2*lambda*F*(M/(2*(M + H)) + (1 - m)*H/(M + H)) - (mu + gamma*N)*F
    )
    
    Mn <- M + tau * (D/h^2 * (Mp - 2*M + Mm) + 2*lambda*F*(M/(2*(M + H))) - (mu + gamma*N)*M
    )
    
  wH[ ,j+1] <- signif(Hn,4)
  wF[ ,j+1] <- signif(Fn,4)
  wM[ ,j+1] <- signif(Mn,4)
  }
  list(x=xLong, t=t, H=wH, F=wF, M=wM)
}
```

## Section 6 - Plotting the Figures

To plot the results I need to load the package `plot3Drgl`.
```{r, Load Plot Packages, echo = FALSE}
library("plot3Drgl")
```

### Plotting the individual populations

I shall now plot the results. Here is the population density of the driving-Y population:
```{r, Plotting the Driving-Y Population}
persp3D(xLong/a, t, wH/N0,
        xlab = "x", ylab = "t", zlab = "H/N0",
        ticktype = "detailed", nticks = 4,
        main = "Population Density of the Driving-Y Population across the Vector Space")
plotrgl(lighting = TRUE, smooth = TRUE)
```

Here is the population density of the female population:
```{r, Plotting the Female Population}
persp3D(xLong/a, t, wF/N0,
        xlab = "x", ylab = "t", zlab = "F/N0",
        ticktype = "detailed", nticks = 4,
        main = "Population Density of the Female Population across the Vector Space")
plotrgl(lighting = TRUE, smooth = TRUE)
```

Here is the population density of the male population:
```{r, Plotting the Male Population}
persp3D(xLong/a, t, wM/N0,
        xlab = "x", ylab = "t", zlab = "M/N0",
        ticktype = "detailed", nticks = 4,
        main = "Population Density of the Male Population across the Vector Space")
plotrgl(lighting = TRUE, smooth = TRUE)
```


### Reproducing graphs from the paper

Below I attempt to reproduce Figure 1 from the paper.
```{r, Reproducing Figure 1}
plot(xLong, (wF[, 501])/N0, type="l", xlab="Dimensionless distance from origin", ylab="Population density", col="red",
     main = "Population densities of the driving-Y, Females and Males")
lines(xLong, (wM[, 501])/N0, type="l", col="black")
lines(xLong, (wH[, 501])/N0, type="l", col="blue")
legend("topleft", legend=c("H", "F", "M"), col=c("blue", "red", "black"), lty=1)
```
There are some differences between the above plot and Figure 1 in the paper however the general shape seems to fit.

Next, I will try to reproduce Figure 2 from the paper showing the affect of changing $m$ on population density of $H$. I will do this by utilising the function `genedrive` that I created in Section 5.

```{r, Reproducing Figure 2}
m75 <- genedrive(m=0.75)
m85 <- genedrive(m=0.85)
m95 <- genedrive(m=0.95)
m100 <- genedrive(m=1.00)

plot((m75$x[151:300])/a, (m75$H[151:300,501])/N0, type="l", xlab="Dimensionless distance from origin", ylab="Population density", col="black", main = "Population densities of the driving-Y, Females and Males",  ylim=c(0, 55))
lines((m85$x[151:300])/a, (m85$H[151:300,501])/N0, m85$H, type="l", col="purple")
lines((m95$x[151:300])/a, (m95$H[151:300,501])/N0, m95$H, type="l", col="blue")
lines((m100$x[151:300])/a, (m100$H[151:300,501])/N0, m100$H, type="l", col="red")
legend("topright", title="m", legend=c("0.75", "0.85", "0.95", "1.00"), col=c("black", "purple", "blue", "red"), lty=1)
```
As you can see, the above graph is very different to Figure 2 in the paper.

## Section 7 - Conclusion

Here, I will discuss the downfalls to my method and the paper and possible improvements I could make.

### Issues with my Method
I implemented the basic forward difference method for my system of equations. From the results I have I think it was a suitable method. The major issues of my method was implementing the $\vec{H_j}^+$ and $\vec{H_j}^-$ since I am not certain they were correct. Despite that I think the results were quite similar to that of the paper. Also my arrangement of the space grid may have caused an issue in the method though it is not apparent in the figures.

### Issues in the Paper
One of the main issues in the paper is that they did not give boundary conditions. Without the boundary conditions they used I could not achieve the correct shape for the driving-Y population. This meant that their results were not reproducible (refer to figure 2). Also their equation for $N_0$ always equaled to $\infty$ which meant that there would have been no plots for females and males were possible so I just took out the $\gamma$ so that it was a positive value.

### Possible Improvements
To improve I would implement the Method of Lines since that was the method used by the authors. I would also vary the boundary conditions to try to fit the shape of the H curve.